### PR TITLE
[PWGHF] Fixed bug in Resonances Task

### PR DESCRIPTION
**Pull request by Luca610**
This PR fixes some bugs in the  TaskCharmResoReduced and changes the computation of the invariant mass in the candidateCreatorCharmResoReduced

### CODE REVIEW

File: PWGHF/D2H/TableProducer/candidateCreatorCharmResoReduced.cxx [hunk 1]
```diff
@@ -328,12 +333,16 @@ struct HfCandidateCreatorCharmResoReduced {
...
+                invMassReso = RecoDecay::m(std::array{candD.pVectorProng0(), candD.pVectorProng1(), candD.pVectorProng2(), pVecV0Tr}, std::array{massPion, massKaon, massPion, massK0}) - invMassD;
+              } else {
+                invMassReso = RecoDecay::m(std::array{candD.pVectorProng1(), candD.pVectorProng0(), candD.pVectorProng2(), pVecV0Tr}, std::array{massPion, massKaon, massPion, massK0}) - invMassD;
+              }
               registry.fill(HIST("hMassDs1"), invMassReso, ptReso);
               break;
             case DecayChannel::Ds2StarToDplusK0s:
               invMassV0 = candV0Tr.invMassK0s();
-              invMassReso = RecoDecay::m(std::array{pVecD, pVecV0Tr}, std::array{massDplus, massK0});
+              invMassReso = RecoDecay::m(std::array{candD.pVectorProng0(), candD.pVectorProng1(), candD.pVectorProng2(), pVecV0Tr}, std::array{massPion, massKaon, massPion, massK0}) - invMassD;
```
Comment (vkucera) [inline PWGHF/D2H/TableProducer/candidateCreatorCharmResoReduced.cxx:345]: Please implement my suggestion from https://github.com/AliceO2Group/O2Physics/pull/8427#discussion_r1841691667

File: PWGHF/D2H/TableProducer/candidateCreatorCharmResoReduced.cxx [hunk 2]
```diff
@@ -172,6 +175,8 @@ struct HfCandidateCreatorCharmResoReduced {
     massDplus = o2::constants::physics::MassDPlus;
     massDstar = o2::constants::physics::MassDStar;
     massD0 = o2::constants::physics::MassD0;
+    massPion = o2::constants::physics::MassPiPlus;
+    massKaon = o2::constants::physics::MassKPlus;
```
Comment (vkucera) [inline PWGHF/D2H/TableProducer/candidateCreatorCharmResoReduced.cxx:179]: This is useless. Just add `using namespace o2::constants::physics;` and use the constants directly.

**Review by vkucera — CHANGES_REQUESTED** [review]

**Comment by Luca610** [comment]
Hi @vkucera, thanks for your review! I impelmented your suggestions in my latest commit

File: PWGHF/D2H/TableProducer/candidateCreatorCharmResoReduced.cxx [hunk 3]
```diff
@@ -315,25 +298,30 @@ struct HfCandidateCreatorCharmResoReduced {
...
+        std::array<std::array<float, 3>, 3> pVectorCharmProngs = {candD.pVectorProng0(), candD.pVectorProng1(), candD.pVectorProng2()};
         float ptReso = RecoDecay::pt(RecoDecay::sumOfVec(pVecV0Tr, pVecD));
 
         if constexpr (channel == DecayChannel::DstarTrack) {
           if (candD.dType() > 0) {
-            invMassReso = RecoDecay::m(std::array{candD.pVectorProng0(), candD.pVectorProng1(), candD.pVectorProng2(), pVecV0Tr}, std::array{massPion, massKaon, massPion, massProton});
+            invMassReso = RecoDecay::m(std::array{pVectorCharmProngs[0], pVectorCharmProngs[1], pVectorCharmProngs[2], pVecV0Tr}, std::array{MassPiPlus, MassKPlus, MassPiPlus, MassProton});
           } else {
-            invMassReso = RecoDecay::m(std::array{candD.pVectorProng1(), candD.pVectorProng0(), candD.pVectorProng2(), pVecV0Tr}, std::array{massPion, massKaon, massPion, massProton});
+            invMassReso = RecoDecay::m(std::array{pVectorCharmProngs[1], pVectorCharmProngs[0], pVectorCharmProngs[2], pVecV0Tr}, std::array{MassPiPlus, MassKPlus, MassPiPlus, MassProton});
```
[the remark this answers is not in the document]
Comment (Luca610) [reply PWGHF/D2H/TableProducer/candidateCreatorCharmResoReduced.cxx:308]: Isn't it the same?

File: PWGHF/D2H/TableProducer/candidateCreatorCharmResoReduced.cxx [hunk 4]
```diff
@@ -315,25 +298,30 @@ struct HfCandidateCreatorCharmResoReduced {
...
+                invMassReso = RecoDecay::m(std::array{pVectorCharmProngs[0], pVectorCharmProngs[1], pVectorCharmProngs[2], pVecV0Tr}, std::array{MassPiPlus, MassKPlus, MassPiPlus, MassK0Short}) - invMassD;
+              } else {
+                invMassReso = RecoDecay::m(std::array{pVectorCharmProngs[1], pVectorCharmProngs[0], pVectorCharmProngs[2], pVecV0Tr}, std::array{MassPiPlus, MassKPlus, MassPiPlus, MassK0Short}) - invMassD;
+              }
               registry.fill(HIST("hMassDs1"), invMassReso, ptReso);
               break;
             case DecayChannel::Ds2StarToDplusK0s:
               invMassV0 = candV0Tr.invMassK0s();
-              invMassReso = RecoDecay::m(std::array{pVecD, pVecV0Tr}, std::array{massDplus, massK0});
+              invMassReso = RecoDecay::m(std::array{pVectorCharmProngs[0], pVectorCharmProngs[1], pVectorCharmProngs[2], pVecV0Tr}, std::array{MassPiPlus, MassKPlus, MassPiPlus, MassK0Short}) - invMassD;
```
[the remark this answers is not in the document]
Comment (Luca610) [reply PWGHF/D2H/TableProducer/candidateCreatorCharmResoReduced.cxx:324]: Is that plus the pVector of the Track/V0

File: PWGHF/D2H/TableProducer/candidateCreatorCharmResoReduced.cxx (hunk 3 above)
[the remark this answers is not in the document]
Comment (vkucera) [reply PWGHF/D2H/TableProducer/candidateCreatorCharmResoReduced.cxx:308]: Not at all.
`std::array{pVectorCharmProngs[0], pVectorCharmProngs[1], pVectorCharmProngs[2]` vs `pVectorCharmProngs` plus you create a new array every time.

File: PWGHF/D2H/TableProducer/candidateCreatorCharmResoReduced.cxx (hunk 4 above)
[the remark this answers is not in the document]
Comment (vkucera) [reply PWGHF/D2H/TableProducer/candidateCreatorCharmResoReduced.cxx:324]: Ah, you have 4 momenta there, sorry, I overlooked it.

File: PWGHF/D2H/TableProducer/candidateCreatorCharmResoReduced.cxx (hunk 3 above)
[the remark this answers is not in the document]
Comment (vkucera) [reply PWGHF/D2H/TableProducer/candidateCreatorCharmResoReduced.cxx:308]: Sorry, I overlooked that you have 4 momenta in the array.

**Review by vkucera — APPROVED** [review]

**Pull request merged**